File: web/apps/cast/src/types/collection.ts
import { EnteFile } from "types/file";
import {
    EncryptedMagicMetadata,
    MagicMetadataCore,
    SUB_TYPE,
    VISIBILITY_STATE,
} from "types/magicMetadata";

export enum COLLECTION_ROLE {
    VIEWER = "VIEWER",
    OWNER = "OWNER",
    COLLABORATOR = "COLLABORATOR",
    UNKNOWN = "UNKNOWN",
}

export interface CollectionUser {
    id: number;
    email: string;
    role: COLLECTION_ROLE;
}

enum CollectionType {
    folder = "folder",
    favorites = "favorites",
    album = "album",
    uncategorized = "uncategorized",
}

export interface EncryptedCollection {
    id: number;
    owner: CollectionUser;
    // collection name was unencrypted in the past, so we need to keep it as optional
    name?: string;
    encryptedKey: string;
    keyDecryptionNonce: string;
    encryptedName: string;
    nameDecryptionNonce: string;
    type: CollectionType;
    attributes: collectionAttributes;
    sharees: CollectionUser[];
    publicURLs?: unknown;
    updationTime: number;
    isDeleted: boolean;
    magicMetadata: EncryptedMagicMetadata;
    pubMagicMetadata: EncryptedMagicMetadata;
    sharedMagicMetadata: EncryptedMagicMetadata;
}

export interface Collection
    extends Omit<
        EncryptedCollection,
        | "encryptedKey"
        | "keyDecryptionNonce"
        | "encryptedName"
        | "nameDecryptionNonce"
        | "magicMetadata"
        | "pubMagicMetadata"
        | "sharedMagicMetadata"
    > {
    key: string;
    name: string;
    magicMetadata: CollectionMagicMetadata;
    pubMagicMetadata: CollectionPublicMagicMetadata;
    sharedMagicMetadata: CollectionShareeMagicMetadata;
}

// define a method on Collection interface to return the sync key as collection.id-time
// this is used to store the last sync time of a collection in local storage

export interface collectionAttributes {
    encryptedPath?: string;
    pathDecryptionNonce?: string;
}

export type CollectionToFileMap = Map<number, EnteFile>;

export interface CollectionMagicMetadataProps {
    visibility?: VISIBILITY_STATE;
    subType?: SUB_TYPE;
    order?: number;
}

export type CollectionMagicMetadata =
    MagicMetadataCore<CollectionMagicMetadataProps>;

export interface CollectionShareeMetadataProps {
    visibility?: VISIBILITY_STATE;
}
export type CollectionShareeMagicMetadata =
    MagicMetadataCore<CollectionShareeMetadataProps>;

export interface CollectionPublicMagicMetadataProps {
    asc?: boolean;
    coverID?: number;
}

export type CollectionPublicMagicMetadata =
    MagicMetadataCore<CollectionPublicMagicMetadataProps>;

export type CollectionFilesCount = Map<number, number>;

File: rust/cli/src/models/collection.rs
#![allow(dead_code)]

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Collection {
    pub id: i64,
    pub owner: i64,
    pub key: String,
    pub name: String,
    #[serde(rename = "type")]
    pub collection_type: CollectionType,
    pub attributes: Option<CollectionAttributes>,
    #[serde(rename = "sharees")]
    pub sharees: Option<Vec<Sharee>>,
    #[serde(rename = "publicURLs")]
    pub public_urls: Option<Vec<PublicURL>>,
    #[serde(rename = "updatedAt")]
    pub updated_at: i64,
    #[serde(rename = "isDeleted")]
    pub is_deleted: bool,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum CollectionType {
    #[serde(rename = "folder")]
    Folder,
    #[serde(rename = "favorites")]
    Favorites,
    #[serde(rename = "album")]
    Album,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CollectionAttributes {
    #[serde(rename = "encryptedPath")]
    pub encrypted_path: Option<String>,
    #[serde(rename = "pathDecryptionNonce")]
    pub path_decryption_nonce: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Sharee {
    pub id: i64,
    pub email: String,
    pub role: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PublicURL {
    pub url: String,
    #[serde(rename = "deviceLimit")]
    pub device_limit: i32,
    #[serde(rename = "validTill")]
    pub valid_till: i64,
}
